service/header/store.go

```diff
@@ -191,28 +241,71 @@ func (s *store) Append(ctx context.Context, headers ...*ExtendedHeader) error {
 		verified, head = append(verified, h), h
 	}
 
-	err = s.put(verified...)
-	if err != nil {
+	// queue headers to be written on disk
+	select {
+	case s.writes <- verified:
+		// once we sure the 'write' is enqueued - update caches with verified headers
+		for _, h := range verified {
+			s.cache.Add(h.Hash().String(), h)
+			s.heightIndex.cache.Add(uint64(h.Height), h.Hash())
+		}
+		// and notify waiters if any + increase current height
+		// it is important to do Pub after updating caches
+		// so cache is consistent with atomic Height counter on the heightSub
+		s.heightSub.Pub(verified...)
+		// we return an error here after writing,
+		// as there might be an invalid header in between of a given range
 		return err
+	case <-s.writesDn:
+		return errStoppedStore
+	case <-ctx.Done():
+		return ctx.Err()
 	}
+}
 
-	err = s.newHead(head.Hash())
-	if err != nil {
-		return err
-	}
+// flushLoop performs writing task to the underlying datastore in a separate routine
+// This way writes are controlled and manageable from one place allowing
+// (1) Appends not to be blocked on long disk IO writes and underlying DB compactions
+// (2) Batching header writes
+func (s *store) flushLoop() {
+	defer close(s.writesDn)
+	for headers := range s.writes {
+		if headers != nil {
+			s.pending = append(s.pending, headers...)
+		}
+		// ensure we flush only if pending is grown enough, or we are stopping(headers == nil)
+		if len(s.pending) < DefaultWriteBatchSize && headers != nil {
+			continue
+		}
 
-	log.Infow("new head", "height", head.Height, "hash", head.Hash())
-	s.heightSub.Pub(verified...)
-	return nil
+		err := s.flush(s.pending...)
+		if err != nil {
+			// TODO(@Wondertan): Should this be a fatal error case with os.Exit?
+			from, to := uint64(headers[0].Height), uint64(headers[len(headers)-1].Height)
+			log.Errorw("writing header batch", "from", from, "to", to)
+		}
+		// reset pending
+		s.pending = s.pending[:0]
+
+		if headers == nil {
```

Comment:
so if the writes queue ever ends up empty for whatever reason, `flushLoop` will exit and never be called at any point again?

Reply:
But could there be a case where `writes` queue can be empty but the node is still technically syncing / trying to stay up to head with the network? What if there's a network issue?

Reply:
What do you mean by empty writes queue? A nil case?

Reply:
Look at append. It will only send non empty values to the writed queue

Reply:
No i mean what happens if all `writes` have been flushed and no new `writes` have been added even though the node is still operational?

Reply:
Nothing happens. Nothing to flush, nothing to writes, so goroutine just blocked on `writes` chan.

Reply:
Not sure if this is resolved. Happy to discuss if there is still something unclear.

Reply:
This is utACKed but I do want to have a call about it later at some point because I don't quite understand how it's not possible in practice that a node could (due to network issues) not receive any new headers from the network and then have writes queue flushed out in the process. It seems like, even though it's unlikely, it's still a possibility. I'm fine w merging as is now but can we talk about it later and maybe write an issue if you see merit in my concern?

Reply:
@renaynay, If there is no writes being enqueue, the pending cache won't grow beyond `DefaultWriteBatchSize` and no flush will happen. This is true, but there is no problem with this as:
* Those headers are still available in the Store cache and anyone can you use them without accessing disk
* The most important one, If the Store is stopped than the flush is mandatory(this is why there is a special case for nil headers) even when pending cache is not big enough, so there won't be case that we miss a flush or some headers.
> then have writes queue flushed out in the process

Reply:
okay so reading `headers := range writes` will never come up empty? My concern not with the flushing, but with accidental exit of the flushLoop with no way to kickstart it again unless node restart.

Reply:
That's the only question i have ^

Reply:
There are no accidental exits of the flushLoop, but purposeful, by the Stop only.

Reply:
Note that I didn't aimed for restartability in this case and there might be issues if Store is getting restarted with Start and Stop multiple times during runtime. I don't see any reason for Store to be restarted tbh, but making this possible would not be hard.